### DISCUSSION TITLE
delete the WebCall user ID

### DISCUSSION THREAD

**Discussion by NCHU611A** [mehrvarz/webcall#24] [Q&A] [answered]
Is there a function to delete the WebCall user ID?
Thank you.

**Comment by mehrvarz** [comment, marked as answer]
WebCall ID's outdate automatically when not being used for a specified duration (6 month). Why delete them explicitly? To make them available for re-use by another party?

**Reply by nscomputing** [reply]
In available documentation WebCall administration is non-existent, but in server source code we can find code that can be related to different common administrative tasks.   
 
For example in the **httpAdmin.go** there is a range of urlPaths such as: "/dumpregistered", "/dumpblocked","/deluserid"......
Also in **main.go** there are config keywords such as: var adminID = "", var adminEmail = "", var adminLogPath1 = "", var adminLogPath2 = ""

Please help me how WebCall server administration works.

Thanks.

**Reply by mehrvarz** [reply]
**adminEmail**: keyword can be specified, but is not being used anywhere in the code. Early on, the idea was to have the server send you an e-mail notification, in case of an instability. But consider this:
1. Some instability occured but could not be detected in code, so no e-mail is being sent.
2. Some instability occured and was detected but because of the issue also no e-mails can be sent.
3. Some instability occured and was detected and e-mail can be sent.

So if you do not receive any automatic e-mail (based on case 3), you can NOT be sure that everything is in order. Solution: it is much better to test the server from the outside (periodically), for instance via simulated clients. This is why "adminEmail" was dropped at some point.

"**/dumpregistered**", "/dumpblocked", "/deluserid" and "/dumponline": These are http handlers that can only be called from localhost. (You SSH into the server and then use curl to invoke.) "**/dumponline**" is the only handler I am using myself. I do so via a script like this:

```
# dumponline.sh
setterm -linewrap off
curl -s http://127.0.0.1:8067/rtcsig/dumponline
setterm -linewrap on
```
The others handlers do still exist, but since account-mgmt has become automatic, they have little utility now. There is just no need to look at user id's anymore.

**adminLogPath1/2** implement functionality similar to /dumponline but over https. I am not yet certain this will stay. If you don't enable this in config.ini, it can not cause issues.

**adminID** can be used to request (long but harmless) thread-traces in the server console. Or in syslog. This is not a "common administrative task" but may be useful for software developers. This feature can be used to detect thread leakage. Right now there are no threads leaking. And if adminID is not configured, this feature cannot be invoked. (I have this enabled in my config.ini on timur.mobi currently, so I can use it should there be a need.)

All the things you refer to are mostly temporary or work-in-progress attempts. Over time it often becomes clear that the first implementation is not sufficient and requres essential changes. It is not unusual to find such work-in-progress traces in code. However, none of this is enabled by default. There is a good chance most of these undocumented things will get removed over time.

There are no required (but undocumented) administration tasks. You should be able to run the service without it needing any additional work. As usual, most work will go into preventing bad actors from kicking your box.

**Reply by nscomputing** [reply]
Thank you very much.

You wrote:   "...most work will go into preventing bad actors from kicking your box."

But, how to do it when, as you wrote: "There are no required (but undocumented) administration tasks. You should be able to run the service without it needing any additional work."?

Looks like that at the moment anyone can kick my box and I can do nothing if I can't easily administer user registration, listings, deleting, blocking.....

**Reply by mehrvarz** [reply]
> I can do nothing if I can't easily administer registration, listings, deleting...

I don't agree. You do not know who is behind a user-ID (unless it's your own or one of a friend). A working user-ID does not provide a bad actor with the ability to apply harm. Why would you want to delete user-ID's and what would be your criteria?

Most if not all of your admin work will go into securing your server box _independent_ of whether WebCall is installed on it. Running WebCall on top will not add much additional work. This is in part because WebCall user-mgmt is automatic. Users can create their ID's themselves. User IDs get deactivated automatically after being idle for six month. When this happens the ID goes into blocked state for two more month, so it can not be re-registered by anyone for at least 8 month. (Most likely it will not get registered for much longer, if it ever gets registered again.) Being able to offer a service that is useful but does not required much labor is part of the concept. Also for my own benefit.

The localhost handlers for register and delete are from the early days of this project, from befor a Web-interface existed for this purpose. I implemented these localhost handlers so I could create the first user accounts in order to start making phone calls. These handlers are now obsolete. They are still in the codebase, but only because you don't rip out working code the moment you discover that it can go. The /dumponline handler will stay. But the other handlers may get removed at any time.

Please let me know if you still have doubts.

**Reply by nscomputing** [reply]
Thank you for your detailed and clear explanation.

**Reply by nscomputing** [reply]
> Please let me know if you still have doubts.

Just one additional question: 

The WebCall ID password is stored in rtchashedpw.db.
 Is it a clear text password or encrypted password?

Thank you.

**Reply by mehrvarz** [reply]
The pw is currently not being encrypted in the db. Here my thinking: We do not deal with money. We do not store any personal data (real name, email address, etc.) anywhere. The only thing a bad actor could do with your pw, is to try to receive your calls and then try to impersonate you. This is rather little payback for a high-skill technical attack.

But in case we want to implement pw-encryption, I have marked the two places in the source code where this would need to go. Proposals welcome:
[httpLogin.go](https://github.com/mehrvarz/webcall/blob/faa15447cf0ee9afcc17603a3d6f928c5d276e25/httpLogin.go#L613) // here we could encrypt the pw
[httpServer.go](https://github.com/mehrvarz/webcall/blob/faa15447cf0ee9afcc17603a3d6f928c5d276e25/httpServer.go#L579) // here we could decrypt the pw

**Reply by nscomputing** [reply]
> We do not store any personal data (real name, email address, etc.) anywhere. 

In my opinion, the personal password (a part of personal data) should be encrypted on the server side. A server administrator must not be in position to compromise account integrity and be able to impersonate an account holder.

> Proposals welcome:
> [httpLogin.go](https://github.com/mehrvarz/webcall/blob/faa15447cf0ee9afcc17603a3d6f928c5d276e25/httpLogin.go#L613) // here we could encrypt the pw
> [httpServer.go](https://github.com/mehrvarz/webcall/blob/faa15447cf0ee9afcc17603a3d6f928c5d276e25/httpServer.go#L579) // here we could decrypt the pw

Please implement saving encrypted password to required database files on the WebCall server. 

My proposal:

1. There is no need to decrypt the pw.
2. Encrypt the password by registration of WebCall ID and save it to required database file(s).
3. Encrypt the password entered by login and compare it to the encrypted password in database file.

Thank you.

**Reply by mehrvarz** [reply]
> A server administrator must not be in position to compromise account integrity and be able to impersonate an account holder.

Administrators will always be in a position to compromise account integrity. How long do you think would it take an ill-minded admin to compromise the most advanced pw-encryption measure? 5 minutes? I can think of no true solution to this. And I am not going to participate in security theater.

But why don't you contribute an actual solution we can look at? Or maybe someone else? I am abs willing to integrate any sufficient solution.

One goal of mine is to not ever show, print or log any user passwords in any situation. So that well-meaning operators do not unintentionally get in touch with them. I personally have not seen a single password in the clear just once since I started this project. (To casually point this out.)

**Reply by nscomputing** [reply]
> But why don't you contribute an actual solution we can look at? Or maybe someone else? I am abs willing to integrate any sufficient solution.

Unfortunately, personally  I am not a Golang developer.

**Reply by mehrvarz** [reply]
> My proposal:
> 
> 1. There is no need to decrypt the pw.
> 2. Encrypt the password by registration of WebCall ID and save it to required database file(s).
> 3. Encrypt the password entered by login and compare it to the encrypted password in database file.

This has been implemented now with some guidence from: https://medium.com/@jcox250/password-hash-salt-using-golang-b041dc94cb72
The new implementation is already online and active for new accounts being registered. If you set up a new server, all passwords will be stored in an encrypted format. I am thinking about converting existing (live) accounts. But there are some operational risks involved with that.

**Reply by mehrvarz** [reply]
Created (one-off) code to encrypt all remaining passwords. Ran this code today without a single clients noticing. Now there are zero unencrypted passwords left in the WebCall db. Voilà.

**Comment by NCHU611A** [comment]
> WebCall ID's outdate automatically when not being used for a specified duration (6 month).

Got it.

> Why delete them explicitly? To make them available for re-use by another party?

I thought that if new IDs are registered all the time, but there is no deletion mechanism, will it cause insufficient memory or insufficient IDs?

Thank you!

**Reply by mehrvarz** [reply]
> will it cause insufficient memory or insufficient IDs?

Registered callees are kept on disk, not in memory. Callees use up very little disk space, so you really don't have to think about it much. And once you run the service for more than 6 month, old/unused callee entries will get discarted automatically in the background. So from that moment forward the db will only grow if you see more new callee users than old/unused accounts being removed.

The ID range allows every human on earth to have 10 ID's per server. Or if 'only' 1 in 10 humans use the service, they can each have 100 ID's, etc. There are two actual limiting factors: 1. number of concurrent/online callees (roughly 30K per GB RAM); 2. concurrent call-requests (roughly 300 per min per intel/amd-core, or up to 440K per 24h per core, probably more). I think an "Advanced server topics" page may be needed, like the one that exists already for the Android client.

**Reply by NCHU611A** [reply]
Got it.
Thank you!
